File: backend/app/api/routes/agents.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.orm import Session

from app.api.deps.auth import get_current_user
from app.core.settings import Settings, get_settings
from app.db.session import get_db_session
from app.models.user import User
from app.models.user_agent import UserAgent
from app.repositories.agent_repository import AgentRepository
from app.repositories.model_provider_settings_repository import ModelProviderSettingsRepository
from app.schemas.agent import (
    AgentIconResponse,
    AgentGenerationStatusResponse,
    UserAgentCreateRequest,
    UserAgentGenerateRequest,
    UserAgentResponse,
    UserAgentsListResponse,
    UserAgentUpdateRequest,
)
from app.services.agent_service import AgentService
from app.services.exceptions import ServiceError

router = APIRouter(prefix="/agents", tags=["agents"])


def _get_agent_service(session: Session) -> AgentService:
    return AgentService(
        session=session,
        agent_repository=AgentRepository(),
        model_provider_settings_repository=ModelProviderSettingsRepository(),
    )


def to_user_agent_response(agent: UserAgent) -> UserAgentResponse:
    return UserAgentResponse(
        id=agent.id,
        name=agent.name,
        variant=agent.icon_key or "empty",
        system_prompt=agent.system_prompt,
        is_custom=agent.default_key is None,
        is_empty=False,
        created_at=agent.created_at,
        updated_at=agent.updated_at,
    )


@router.get("/icons", response_model=list[AgentIconResponse])
def list_agent_icons(
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
) -> list[AgentIconResponse]:
    del current_user
    service = _get_agent_service(session)
    return [AgentIconResponse.model_validate(icon) for icon in service.list_icons()]


@router.get("", response_model=UserAgentsListResponse)
def list_agents(
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
) -> UserAgentsListResponse:
    service = _get_agent_service(session)
    agents = service.list_agents(user=current_user)
    return UserAgentsListResponse(items=[to_user_agent_response(agent) for agent in agents])


@router.get("/generation-status", response_model=AgentGenerationStatusResponse)
def get_agent_generation_status(
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
    settings: Settings = Depends(get_settings),
) -> AgentGenerationStatusResponse:
    del current_user
    service = _get_agent_service(session)
    return AgentGenerationStatusResponse(available=service.can_generate_agent_prompt(settings=settings))


@router.post("", response_model=UserAgentResponse, status_code=status.HTTP_201_CREATED)
def create_agent(
    payload: UserAgentCreateRequest,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
    settings: Settings = Depends(get_settings),
) -> UserAgentResponse:
    service = _get_agent_service(session)
    try:
        agent = service.create_agent(
            user=current_user,
            name=payload.name,
            icon_key=payload.variant,
            system_prompt=payload.system_prompt,
            max_agents=settings.user_max_agents,
        )
    except ServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.detail) from exc
    return to_user_agent_response(agent)


@router.patch("/{agent_id}", response_model=UserAgentResponse)
def update_agent(
    agent_id: uuid.UUID,
    payload: UserAgentUpdateRequest,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
) -> UserAgentResponse:
    service = _get_agent_service(session)
    try:
        agent = service.update_agent(
            user=current_user,
            agent_id=agent_id,
            name=payload.name,
            icon_key=payload.variant,
            system_prompt=payload.system_prompt,
        )
    except ServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.detail) from exc
    return to_user_agent_response(agent)


@router.post("/{agent_id}/generate", response_model=UserAgentResponse)
def generate_agent(
    agent_id: uuid.UUID,
    payload: UserAgentGenerateRequest,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
    settings: Settings = Depends(get_settings),
) -> UserAgentResponse:
    service = _get_agent_service(session)
    try:
        agent = service.generate_agent(
            user=current_user,
            agent_id=agent_id,
            name=payload.name,
            system_prompt=payload.system_prompt,
            settings=settings,
        )
    except ServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.detail) from exc
    return to_user_agent_response(agent)


@router.delete("/{agent_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_agent(
    agent_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_db_session),
) -> Response:
    service = _get_agent_service(session)
    try:
        service.delete_agent(user=current_user, agent_id=agent_id)
    except ServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.detail) from exc
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/features/workspace/api/agents.js
import { getApiBaseUrl } from "../../../api/baseUrl";
import { readWorkspaceApiError } from "./readWorkspaceApiError";

function mapAgent(dto) {
  return {
    id: dto.id,
    name: dto.name,
    variant: dto.variant ?? "empty",
    systemPrompt: dto.system_prompt ?? "",
    isCustom: Boolean(dto.is_custom),
    isEmpty: Boolean(dto.is_empty),
    isPendingSetup: false,
    createdAt: dto.created_at ?? null,
    updatedAt: dto.updated_at ?? null,
  };
}

function mapSelectedAgent(dto) {
  return {
    ...mapAgent(dto),
    placement: dto.placement,
    position: dto.position,
  };
}

function mapVariantToApi(variant) {
  return variant && variant !== "empty" ? variant : null;
}

async function readJsonResponse(response, fallbackMessage) {
  if (!response.ok) {
    const detail = await readWorkspaceApiError(response, fallbackMessage);
    throw new Error(detail);
  }

  return response.json();
}

export async function listAgents({ accessToken, signal }) {
  const response = await fetch(`${getApiBaseUrl()}/agents`, {
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    signal,
  });
  const payload = await readJsonResponse(response, "Не удалось загрузить агентов");

  return (payload.items ?? []).map(mapAgent);
}

export async function getAgentGenerationStatus({ accessToken, signal }) {
  const response = await fetch(`${getApiBaseUrl()}/agents/generation-status`, {
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    signal,
  });
  const payload = await readJsonResponse(response, "Не удалось проверить доступность генерации");

  return Boolean(payload.available);
}

export async function createAgent({ accessToken, name, variant, systemPrompt }) {
  const response = await fetch(`${getApiBaseUrl()}/agents`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      name,
      variant: mapVariantToApi(variant),
      system_prompt: systemPrompt,
    }),
  });
  const payload = await readJsonResponse(response, "Не удалось создать агента");

  return mapAgent(payload);
}

export async function updateAgent({ accessToken, agentId, name, variant, systemPrompt }) {
  const response = await fetch(`${getApiBaseUrl()}/agents/${agentId}`, {
    method: "PATCH",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      name,
      variant: mapVariantToApi(variant),
      system_prompt: systemPrompt,
    }),
  });
  const payload = await readJsonResponse(response, "Не удалось сохранить агента");

  return mapAgent(payload);
}

export async function deleteAgent({ accessToken, agentId }) {
  const response = await fetch(`${getApiBaseUrl()}/agents/${agentId}`, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  });

  if (!response.ok) {
    const detail = await readWorkspaceApiError(response, "Не удалось удалить агента");
    throw new Error(detail);
  }
}

export async function generateAgent({ accessToken, agentId, name, systemPrompt }) {
  const response = await fetch(`${getApiBaseUrl()}/agents/${agentId}/generate`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      name,
      system_prompt: systemPrompt,
    }),
  });
  const payload = await readJsonResponse(response, "Не удалось сгенерировать промпт агента");

  return mapAgent(payload);
}

export async function listChatSessionAgents({ accessToken, chatSessionId, signal }) {
  const response = await fetch(`${getApiBaseUrl()}/chat-sessions/${chatSessionId}/agents`, {
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
    signal,
  });
  const payload = await readJsonResponse(response, "Не удалось загрузить агентов чата");

  return (payload.items ?? []).map(mapSelectedAgent);
}

export async function attachChatSessionAgent({ accessToken, chatSessionId, agentId, placement, position }) {
  const response = await fetch(`${getApiBaseUrl()}/chat-sessions/${chatSessionId}/agents`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      agent_id: agentId,
      placement,
      position,
    }),
  });
  const payload = await readJsonResponse(response, "Не удалось добавить агента в чат");

  return (payload.items ?? []).map(mapSelectedAgent);
}

export async function detachChatSessionAgent({ accessToken, chatSessionId, agentId }) {
  const response = await fetch(`${getApiBaseUrl()}/chat-sessions/${chatSessionId}/agents/${agentId}`, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`,
    },
  });
  const payload = await readJsonResponse(response, "Не удалось убрать агента из чата");

  return (payload.items ?? []).map(mapSelectedAgent);
}
